src/electron/ipc.ts

```diff
@@ -0,0 +1,298 @@
+  [PUSH_CLIPBOARD_CHANNEL]: () => void;
+  [ADD_SERVER_CHANNEL]: (url: string) => void;
+  [VPN_CONNECTED_CHANNEL]: (id: string) => void;
+  [VPN_RECONNECTING_CHANNEL]: (id: string) => void;
+  [VPN_DISCONNECTED_CHANNEL]: (id: string) => void;
+  [REQUEST_LOCALIZATION_CHANNEL]: (localizationKeys: string[]) => void;
+  [LOCALIZATION_RESPONSE_CHANNEL]: (result?: {[key: string]: string}) => void;
+}
+
+/**
+ * Define the contract (request and response type) of all two-way IPC channels.
+ * This interface should only be used for type definitions, please **do not**
+ * try to implement it.
+ *
+ * Please do not include `Promise` or `ErrorCode` in the return type, they will
+ * be automatically handled by `OutlineIpcClient` and `OutlineIpcHandler`.
+ *
+ * We use `ErrorCode` (a.k.a, a `number`) to pass exceptions because catching
+ * custom errors (even as simple as numbers) does not work in electron:
+ *   - https://github.com/electron/electron/issues/24427
+ */
+export interface OutlineIpcTwoWayService {
+  [INSTALL_SERVICES_CHANNEL]: () => void;
+  [CHECK_REACHABLE_CHANNEL]: (hostname: string, port: number) => boolean;
+  [START_VPN_CHANNEL]: (config: ShadowsocksConfig, id: string) => void;
+  [STOP_VPN_CHANNEL]: () => void;
+}
+
+/**
+ * This is a type definition of all one-way IPC channels which can be used as
+ * the type of an IPC message string (`'quit-app' | 'add-server' | ...`).
+ */
+export type OutlineIpcOneWayChannels = typeof OUTLINE_ONE_WAY_IPC_CHANNELS[number] & keyof OutlineIpcOneWayService;
+
+/**
+ * This is a type definition of all two-way IPC channels which can be used as
+ * the type of an IPC message string (`'start-proxying' | ...`).
+ */
+export type OutlineIpcTwoWayChannels = typeof OUTLINE_TWO_WAY_IPC_CHANNELS[number] & keyof OutlineIpcTwoWayService;
+
+/**
+ * The IPC response type across boundaries. This type should be shared between
+ * `.handle` and `.invoke` methods.
+ *
+ * We use `ErrorCode` (a.k.a, a `number`) to pass exceptions because catching
+ * custom errors (even as simple as numbers) does not work in electron:
+ *   - https://github.com/electron/electron/issues/24427
+ */
+export type IpcInvokeResultCrossBoundary<T extends OutlineIpcTwoWayChannels> = Promise<
+  [ErrorCode, ReturnType<OutlineIpcTwoWayService[T]>]
+>;
+
+/**
+ * The IPC sender class which can be used by both main process and renderer
+ * process. The caller is responsible to give it the correct implementation
+ * of the underlying IPC strategy (for example, `mainWindow.webContents`,
+ * or an object exposed by preload script).
+ *
+ * @example
+ *   // In main process
+ *   const client = new OutlineIpcClient(mainWindow.webContents);
+ *   client.send(APP_UPDATE_DOWNLOADED_CHANNEL);
+ *
+ *   // In renderer process (window.ipcImpl is injected by preload)
+ *   const client = new OutlineIpcClient(window.ipcImpl);
+ *   const result = await client.invoke(STOP_VPN_CHANNEL);
+ */
+export class OutlineIpcClient {
```

Comment:
This class and `OutlineIpcPreloadImpl` should probably be merged as a `IpcRendererMethodChannel` implementation.